libp2p/switch.nim

```diff
@@ -206,19 +206,19 @@ proc identify(s: Switch, conn: Connection) {.async, gcsafe.} =
 
     trace "identify: identified remote peer", peer = $conn.peerInfo
 
-proc mux(s: Switch, conn: Connection) {.async, gcsafe.} =
+proc mux(s: Switch, conn: Connection): Future[bool] {.async, gcsafe.} =
   ## mux incoming connection
 
   trace "muxing connection", peer = $conn
   let muxers = toSeq(s.muxers.keys)
   if muxers.len == 0:
     warn "no muxers registered, skipping upgrade flow"
-    return
+    return false
 
   let muxerName = await s.ms.select(conn, muxers)
   if muxerName.len == 0 or muxerName == "na":
```

Comment:
this should also check that muxerName is in s.muxers - ie the relationship is too weak to rely on, that only "supported" muxes will be returned

Reply:
Not sure I understand this one sorry,
we take all muxer names from s.muxers above and pass it to select, that routine will do that check already afair and return ne or empty string (I fixed that recently).

Reply:
it's just a weird way of calling the select - with the names and then looking up in collection, and also checking for empty or "na" - why doesn't ms.select return the muxer instance and that's it or something like that?

Reply:
I think that is outside of the scope of this PR, I absolutely agree tho, but requires a much big refactor.

Reply:
`select` should not return the instance of anything. Multistream-select knows nothing about mixers, secure channels, etc... All it does is signal the other end that a request for a particular string needs to be handled, it is up to the upper layers to decide what to do with the stream once negotiated.

Reply:
> this should also check that muxerName is in s.muxers - ie the relationship is too weak to rely on, that only "supported" muxes will be returned

No this is actually what the protocol expects, it will only return the strings for those ones that passed the negotiation. The relation is fine in the context of a string based protocol negotiation mechanism.

Reply:
well, kind of not really - ie you want the lookup and validation to be close to the resolution of the result - ie you don't want two different modules to be comparing the same string. effectively, `s.muxers[...]` performs a second lookup for the same value, and the two may go out of sync in between as code gets refactored - if instead for example a map was passed in `let muxer = s.ms.select(conn, s.muxers)`, this could be done in  a single pass with lower maintenance risk